# Duplicates in Dataset

## What are Duplicates?
Duplicates are rows in a dataset that have exactly the same values across all columns (or sometimes specific columns). These repeated records do not add new information.

## Why Duplicates are a Problem?
- They can **bias the model** by over-representing certain data.
- They may lead to **incorrect insights** during analysis.
- They can **increase dataset size unnecessarily**, slowing down processing.

## When Do Duplicates Occur?
- Data entry mistakes
- Merging multiple datasets
- System or logging errors
- Web scraping issues

## How to Handle Duplicates?
1. **Detect duplicates** → Identify repeated rows  
2. **Analyze them** → Decide if they are truly unwanted  
3. **Remove or keep** → Based on context and need  

## When Should You Remove Duplicates?
- When duplicates are **unintentional errors**
- When they **distort analysis or model training**

## When Should You Keep Duplicates?
- When duplicates represent **real repeated events** (e.g., transactions, clicks)
- When frequency of occurrence is **important information**

> Not all duplicates are bad — always understand your data before removing them.

---

## Dataset Used

We are using the Titanic dataset available in seaborn

This dataset contains information about passengers such as age, gender, class, and whether they survived or not.

In [4]:
import seaborn as sns
df = sns.load_dataset("titanic")

---

## Checking Number of Duplicate Rows

In [5]:
df.duplicated().sum()

np.int64(107)

### Explanation
- `df.duplicated()` → Checks each row and returns **True** if the row is a duplicate, otherwise **False**
- `.sum()` → Counts all **True values** (since True = 1, False = 0)
- The dataset contains **107 duplicate rows**
- `np.int64` means the result is an integer type from NumPy

### When to Use This?
- At the **initial stage of data cleaning**
- To quickly check if your dataset has duplicate entries

---

## Removing Duplicate Rows

In [6]:
df.drop_duplicates()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,0,3,female,39.0,0,5,29.1250,Q,Third,woman,False,NaN,Queenstown,no,False
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


### Explanation:
- Removes all **duplicate rows** from the dataset  
- Keeps only the **first occurrence** of each duplicate row  
- Returns a **new DataFrame** (original data remains unchanged)  
- The dataset now contains **only unique rows**  
- Duplicate rows have been removed  
- Final shape: **784 rows × 15 columns**  
- This means **107 duplicate rows** were removed from the original dataset  

### When to Use This?
- After detecting duplicates in your dataset  
- When duplicate rows are **errors or unnecessary**  
- Before training ML models to avoid bias  